components/brave_federated/communication_adapter.cc

```diff
@@ -0,0 +1,197 @@
+  auto request = MakeResourceRequest();
+  VLOG(2) << "FL: Posting Task results " << request->method << " "
+          << request->url;
+
+  const std::string payload = BuildUploadTaskResultsPayload(result);
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(request), GetNetworkTrafficAnnotationTag());
+  url_loader_->AttachStringForUpload(payload, "application/protobuf");
+  url_loader_->DownloadToString(
+      url_loader_factory_.get(),
+      base::BindOnce(&CommunicationAdapter::OnUploadTaskResult,
+                     weak_factory_.GetWeakPtr(), std::ref(callback)),
+      kMaxFederatedServerResponseSizeBytes);
+}
+
+void CommunicationAdapter::OnGetTasks(
+    GetTaskCallback& callback,
+    const std::unique_ptr<std::string> response_body) {
+  CHECK(url_loader_);
+  CHECK(url_loader_->ResponseInfo());
+
+  bool failed_request =
+      !url_loader_->ResponseInfo() || !url_loader_->ResponseInfo()->headers;
+  reconnect_backoff_entry_->InformOfRequest(!failed_request);
+  base::TimeDelta reconnect_delay_in_seconds =
+      reconnect_backoff_entry_->GetTimeUntilRelease();
+
+  if (failed_request) {
+    VLOG(2) << "FL: Failed to request tasks, retrying in "
+            << reconnect_delay_in_seconds;
+    return std::move(callback).Run({}, reconnect_delay_in_seconds);
+  }
+
+  auto headers = url_loader_->ResponseInfo()->headers;
+  int response_code = headers->response_code();
+  if (response_code == net::HTTP_OK &&
+      headers->HasHeaderValue("Content-Type", "application/protobuf")) {
+    auto task_list = ParseTaskListFromResponseBody(*response_body);
+
+    request_task_backoff_entry_->InformOfRequest(task_list.has_value());
+    base::TimeDelta request_task_delay_in_seconds =
+        request_task_backoff_entry_->GetTimeUntilRelease();
+    if (!task_list.has_value()) {
+      VLOG(2) << "FL: ParseTaskList failed. Reason: " << task_list.error()
+              << ". Retrying in " << request_task_delay_in_seconds;
+      return std::move(callback).Run({}, request_task_delay_in_seconds);
+    }
+
+    VLOG(2) << "FL: Received " << task_list.value().size()
+            << " tasks from FL service";
+
+    return std::move(callback).Run(task_list.value(),
+                                   request_task_delay_in_seconds);
+  }
+
+  VLOG(2) << "FL: Failed to request tasks. Response code: " << response_code;
+}
+
+void CommunicationAdapter::OnUploadTaskResult(
+    UploadResultCallback& callback,
+    std::unique_ptr<std::string> response_body) {
+  CHECK(url_loader_);
+  CHECK(url_loader_->ResponseInfo());
+
+  if (!url_loader_->ResponseInfo() || !url_loader_->ResponseInfo()->headers) {
+    VLOG(2) << "FL: Failed to post task results";
+    TaskResultResponse response(/*successful*/ false);
+    return std::move(callback).Run(response);
+  }
+
+  auto headers = url_loader_->ResponseInfo()->headers;
+  int response_code = headers->response_code();
+  if (response_code == net::HTTP_OK) {
```

Comment:
nit: As a reader, I expect that we process fail conditions first and handle the success at the end of the function.

So I would process the failure condition here instead:
`  if (response_code != net::HTTP_OK) {`

Reply:
good point. I'll keep that in mind 👍